Add S3DF documentation (#6061)
BLAST-WarpX/warpx: Docs/source/install/hpc/s3df.rst

.. _building-s3df:

> S3DF (SLAC)
> =============

Reviewer: Just some alignment/formatting:
```suggestion
S3DF (SLAC)
===========
```

The S3DF cluster is located at SLAC.

Introduction
------------

If you are new to this system, **please see the following resources**:

> * `S3DF documentation <https://s3df.slac.stanford.edu/#/>`__

Reviewer: Slightly simpler link:
```suggestion
* `S3DF documentation <https://s3df.slac.stanford.edu>`__
```

* Batch system: `Slurm <https://github.com/slaclab/sdf-docs/blob/main/batch-compute.md>`__
* Filesystem locations:

>     * User folder: ``$HOME>`` (25GB on Weka file system)

Reviewer: ```suggestion
    * User folder: ``$HOME`` (25GB on Weka file system)
```

    * Scratch folder: ``/sdf/scratch/<username_initial>/<username>`` (100GB on Weka file system)
    * Group storage: Depends on your group, see e.g. information for the ATLAS group `here <https://usatlas.readthedocs.io/projects/af-docs/en/latest/sshlogin/ssh2SLAC/>`__

Through S3DF we have access to CPU and GPU nodes (the latter equipped with NVIDIA A100).

Installation
------------

> 1) Prerequisites
> ^^^^^^^^^^^^^^^^^^^^^^^^^

Reviewer: ```suggestion
1) Prerequisites
^^^^^^^^^^^^^^^^
```

- S3DF login to a node with access to A100 GPUs (via SLURM job or interactive session).
- Conda available in your shell (Miniconda/Anaconda on S3DF user space).
- GPU: NVIDIA A100 (sm_80).

For other GPUs, adjust ``-DCMAKE_CUDA_ARCHITECTURES``:

- V100: `70`
- A100: `80`
- H100: `90`

.. note::

  SLURM tip (S3DF):

  -  Use srun to have GPU nodes allocated. WarpX requires a "task" per CPU-GPU pair, so if you want to run WarpX e.g. on 4 GPUs, you would do:

  .. code-block:: bash

    srun --pty --cpus-per-task=8 --ntasks-per-node=4 --gpus=4 --mem=100GB --nodes=1 --time=2:00:00 --partition=ampere --account=atlas bash

2) Create & Populate the Conda Environment
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  .. code-block:: bash

    # A) create env (example set; versions can vary by site repos)

>     conda create -n warpx-gpu-test -y python=3.11
>     conda activate warpx-gpu-test

Reviewer: Maybe remove the `-test` suffix?
```suggestion
    conda create -n warpx-gpu -y python=3.11
    conda activate warpx-gpu
```

    #Get the full CUDA toolkit in your conda env.
    conda install -c nvidia -y "cuda=12.9.*"

    # Additionally install NVTX dev tools (cuda separates from the runtime version)
    conda install -c nvidia -y "cuda-nvtx-dev=12.9.*"

    # B) Install build tools
    conda install -c conda-forge -y \
      "cmake>=3.27" ninja make pkg-config git \

>       openmpi ucx mpi4py \

Reviewer: Pin MPI flavor for mpi4py as well?
```suggestion
      openmpi ucx mpi4py=*=mpi_openmpi* \
```

>       gcc_linux-64=13 gxx_linux-64=13

Reviewer: That works, or you can use
```suggestion
      compilers=1.10
```
But maybe your version is more explicit than the meta-package.

>     # C) Install ADIOS2 with OpenMPI (so we can use openpmd_backend = bp)
>     conda install -c conda-forge "adios2=*=mpi_openmpi*"

Reviewer: ADIOS2 is perfect and best performance. If you like for completeness, you can also add hdf5
```suggestion
    # C) Install ADIOS2 and HDF5 with OpenMPI (so we can use openpmd_backend = bp or = h5)
    conda install -c conda-forge "adios2=*=mpi_openmpi* hdf5=*=mpi_openmpi*"
```

    # D) Verify build tools
    which cmake && cmake --version     # should show cmake >= 3.24 from your env
    which nvcc  && nvcc  --version     # should show CUDA 12.9 from the nvidia 'cuda' metapackage
    which mpicxx && ompi_info | head   # Open MPI from conda-forge

.. note::
  - ``cuda-nvtx-dev`` provides the nvtx3 headers used by recent NVTX versions.
  - ``cuda-nvcc`` brings nvcc under ``$CONDA_PREFIX/bin/nvcc``.
  - MPI here is via ``mpich`` in the env (``mpicc``, ``mpicxx`` will be at ``$CONDA_PREFIX/bin/...``).

> 3) Download WarpX
> ^^^^^^^^^^^^^^^^^
> 
> .. code-block:: bash
> 
>   git clone https://github.com/BLAST-WarpX/warpx.git
>   cd warpx

Reviewer 2: I would add a hint as to where this should happen, e.g. in the $HOME folder.

```suggestion
3) Download WarpX
^^^^^^^^^^^^^^^^^

.. code-block:: bash
  cd $HOME
  git clone https://github.com/BLAST-WarpX/warpx.git
  cd warpx
```

4) Environment Fixes (NVTX headers + dlopen/dlsym)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Recent NVTX installs the header under ``nvtx3/nvToolsExt.h``. Add it to your include path, and ensure ``libdl`` is linked (``CUDA/NVTX`` uses ``dlopen()/dlsym()`` dynamically):

Reviewer: Interesting, I do not need these NVTX quirks once the `cuda-nvtx-dev` package is installed:
https://github.com/BLAST-ImpactX/impactx-benchmarks/blob/main/napac25/benchmark-gpu-conda.yaml

.. code-block:: bash

  # ensure libdl is not dropped by the linker (GCC + gold/ld with --as-needed)
  export LDFLAGS="${LDFLAGS} -Wl,--no-as-needed -ldl"

  # verify the header exists (one of these should print a path)
  ls "$CONDA_PREFIX/targets/x86_64-linux/include/nvtx3/nvToolsExt.h" \
   || ls "$CONDA_PREFIX/include/nvtx3/nvToolsExt.h"

  # include & library paths from the conda CUDA toolchain
  export CUDACXX="$CONDA_PREFIX/bin/nvcc"
  export CUDA_HOME="$CONDA_PREFIX"

  export CPATH="$CONDA_PREFIX/targets/x86_64-linux/include/nvtx3:$CONDA_PREFIX/targets/x86_64-linux/include"
  export LIBRARY_PATH="$CONDA_PREFIX/targets/x86_64-linux/lib"
  export LD_LIBRARY_PATH="$CONDA_PREFIX/targets/x86_64-linux/lib"

  # help CMake find ADIOS2/openPMD (if using conda-provided openPMD)
  export ADIOS2_DIR="$CONDA_PREFIX"

>   export openPMD_DIR="$CONDA_PREFIX"

Reviewer: I usually find that `conda activate` sets the `CMAKE_PREFIX_PATH` environment variable, which CMake uses to find all dependencies. Thus, I think the quirks mentioned in section 4) might actually not be needed. Can you try again?

5) Configure with CMake (CUDA + MPI, 3D)

> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Reviewer 2: ```suggestion
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

.. code-block:: bash

>   "$CONDA_PREFIX/bin/cmake" -S . -B build \

Reviewer: Did you run `conda activate`? Could just read
```suggestion
  cmake -S . -B build \
```
then

    -DWarpX_APP=ON \
    -DWarpX_DIMS="3" \
    -DWarpX_COMPUTE=CUDA \
    -DCMAKE_CUDA_ARCHITECTURES=80 \
    -DCMAKE_CUDA_RUNTIME_LIBRARY=Shared \
    -DCMAKE_CUDA_COMPILER="$CONDA_PREFIX/bin/nvcc" \
    -DCMAKE_CUDA_HOST_COMPILER="$(which mpicxx)" \
    -DCMAKE_C_COMPILER="$(which mpicc)" \

>     -DCMAKE_CXX_COMPILER="$(which mpicxx)" \

Reviewer: [I would usually set](https://github.com/BLAST-WarpX/warpx/blob/f7ede6b04756fb95540e94d15fe196953ab00b87/Tools/machines/lawrencium-lbnl/lawrencium_warpx.profile.example#L48-L53), after `conda activate warpx-gpu`
```bash
# optimize CUDA compilation for V100
export AMREX_CUDA_ARCH=7.0
# optimize CUDA compilation for A100
#export AMREX_CUDA_ARCH=8.0
# optimize CUDA compilation for H100
#export AMREX_CUDA_ARCH=9.0

# compiler environment hints
export CXX=$(which g++)
export CC=$(which gcc)
export FC=$(which gfortran)
export CUDACXX=$(which nvcc)
export CUDAHOSTCXX=${CXX}
```

and then CMake picks those up. No need to use the MPI Wrappers as CC/CXX either, because CMake's `find_package(MPI ...)` sets the MPI flags up.

    -DWarpX_FFT=ON

.. note::
  What these do:

  - ``WarpX_APP=ON``: build the warpx application.
  - ``WarpX_DIMS="3"``: build the 3D executable.
  - ``WarpX_COMPUTE=CUDA``: enable GPU backend.
  - ``CMAKE_CUDA_ARCHITECTURES=80``: targets A100.
  - ``CMAKE_CUDA_RUNTIME_LIBRARY=Shared``: link CUDA runtime dynamically (matches conda toolchain layout).
  - Compilers explicitly point to MPI wrappers from the env.

6) Build
^^^^^^^^

.. code-block:: bash

>   "$CONDA_PREFIX/bin/cmake" --build build -j 8

Reviewer 2: Similarly to what @ax3l pointed out above:

```suggestion
  cmake --build build -j 8
```

> 7) Run (MPI + GPU)
> ^^^^^^^^^^^^^^^^^^
> 

Reviewer 2: I would recall that one needs to request the resources first with `srun`, in interactive mode. 

```suggestion
7) Run (MPI + GPU)
^^^^^^^^^^^^^^^^^^

Interactive
"""""""""""
Request resources, for example:

.. code-block:: bash

    srun --pty --nodes=1 --ntasks-per-node=4 --cpus-per-task=8 --gpus=4 --mem=100GB  --time=2:00:00 --partition=ampere --account=<account> bash
```

> From the build dir:
> 
> .. code-block:: bash
> 
>   mkdir run_test  # create run folder for each new test!
>   cp build/bin/warpx.3d run_test/  # copy warpx executable to each run folder!
>   cp Examples/Physics_applications/beam_beam_collision/inputs_test_3d_beam_beam_collision run_test/ #copy input file!
>   cd run_test/
>   mpirun -np 4 ./warpx.3d inputs_test_3d_beam_beam_collision # run!

Reviewer 2: I'm not sure that I would recommend working inside the build directory. 
Usually, the best practice is to run in the scratch folder. 
I'm also not sure about recommending copying the executable every time. 

```suggestion

Move to scratch:

.. code-block:: bash

  cd /sdf/scratch/<first_letter_of_username>/<username>

Create and move into run folder for each new test:

.. code-block:: bash

  mkdir run_test  
  cd run_test 

Copy input file and run:

.. code-block:: bash

  cp $HOME/warpx/Examples/Physics_applications/beam_beam_collision/inputs_test_3d_beam_beam_collision . 
  mpirun -np 4 $HOME/warpx/build/bin/warpx.3d inputs_test_3d_beam_beam_collision 
```

Reviewer 2: You could also add instructions to run in batch mode, if you think it would be useful.
If you need support, we're here :wink:
